### PR TITLE
Update offsets for Skyrim AE 1.6.318

### PR DESCRIPTION
**Pull request by Nukem9**
Title.

I had to switch to REL::Offset because an updated version of the address library is not available. There's a script to easily swap between the two so that can be done later. These were all manually translated with BinDiff or direct code signatures.

NiRTTI and RTTI were generated with [powerofthree's SKSE plugin](https://github.com/powerof3/RTTIDumpSE).

Certain functions or RTTI entries don't exist any longer. I don't know how you want to resolve these:
~~[here](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/compare/master...Nukem9:master#diff-9bd04ac9ffb35ca549191281fa321a798c84b00fb0f496231bc9e4bc466933f6R333)~~
~~[here](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/compare/master...Nukem9:master#diff-9bd04ac9ffb35ca549191281fa321a798c84b00fb0f496231bc9e4bc466933f6R385)~~
~~[here](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/compare/master...Nukem9:master#diff-9bd04ac9ffb35ca549191281fa321a798c84b00fb0f496231bc9e4bc466933f6R387)~~
[here](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/compare/master...Nukem9:master#diff-9bd04ac9ffb35ca549191281fa321a798c84b00fb0f496231bc9e4bc466933f6R429)
[here](https://github.com/Ryan-rsm-McKenzie/CommonLibSSE/compare/master...Nukem9:master#diff-47b26e949ef30ea3b793bf37c3e2325ef2ecb07eb0b31dd99158b32c952cbbe9R17)
[removed](include/RE/L/LoggingDisabler.h)

### CODE REVIEW

**Comment by Exit-9B** [comment]
From what I can tell:
`GFxValue::ObjectInterface::GetArraySize`: 0x00F0C310
`InventoryChanges::GetNextUniqueID`: 0x001F87B0 (another function 0x001F8600 looks similar though)
`LooseFileStream::ctor`: Gone, but there are 2 other constructors 0x00C72BF0 and 0x00C72E70
`ScriptEventSourceHolder::SendActivateEvent`: Inlined

`LoggingDisabler`: Renamed to `DejaGFxLogger`

The only one I've actually tested in a plugin is `GetArraySize`. I can confirm that one works.

File: include/SKSE/Interfaces.h [hunk 1]
```diff
@@ -349,4 +349,31 @@ namespace SKSE
...
+		char          name[256];
+
+		char author[256];
+		char supportEmail[256];
+
+		std::uint32_t versionIndependence;
+		std::uint32_t compatibleVersions[16];
+
+		std::uint32_t seVersionRequired;
+	};
```
Comment (Ryan-rsm-McKenzie) [inline include/SKSE/Interfaces.h:378]:
```suggestion
	struct PluginVersionData
	{
	public:
		enum
		{
			kVersion = 1,
		};

		constexpr void AuthorEmail(std::string_view a_email) noexcept { SetCharBuffer(a_email, std::span{ supportEmail }); }
		constexpr void AuthorName(std::string_view a_name) noexcept { SetCharBuffer(a_name, std::span{ author }); }

		constexpr void CompatibleVersions(std::initializer_list<REL::Version> a_versions) noexcept
		{
			assert(a_versions.size() < std::size(compatibleVersions) - 1);
			std::transform(
				a_versions.begin(),
				a_versions.end(),
				std::begin(compatibleVersions),
				[](const REL::Version& a_version) noexcept { return a_version.pack(); });
		}

		constexpr void MinimumRequiredXSEVersion(REL::Version a_version) noexcept { xseMinimum = a_version.pack(); }
		constexpr void PluginName(std::string_view a_plugin) noexcept { SetCharBuffer(a_plugin, std::span{ pluginName }); }
		constexpr void PluginVersion(REL::Version a_version) noexcept { pluginVersion = a_version.pack(); }
		constexpr void UsesAddressLibrary(bool a_value) noexcept { addressLibrary = a_value; }
		constexpr void UsesSigScanning(bool a_value) noexcept { sigScanning = a_value; }

		const std::uint32_t dataVersion{ kVersion };
		std::uint32_t       pluginVersion = 0;
		char                pluginName[256] = {};
		char                author[256] = {};
		char                supportEmail[256] = {};
		union
		{
			struct
			{
				bool addressLibrary: 1;
				bool sigScanning: 1;
			};
			std::uint32_t padding = 0;
		};
		std::uint32_t compatibleVersions[16] = {};
		std::uint32_t xseMinimum = 0;

	private:
		static constexpr void SetCharBuffer(
			std::string_view a_src,
			std::span<char>  a_dst) noexcept
		{
			assert(a_src.size() < a_dst.size());
			std::fill(a_dst.begin(), a_dst.end(), '\0');
			std::copy(a_src.begin(), a_src.end(), a_dst.begin());
		}
	};
	static_assert(offsetof(PluginVersionData, dataVersion) == 0x000);
	static_assert(offsetof(PluginVersionData, pluginVersion) == 0x004);
	static_assert(offsetof(PluginVersionData, pluginName) == 0x008);
	static_assert(offsetof(PluginVersionData, author) == 0x108);
	static_assert(offsetof(PluginVersionData, supportEmail) == 0x208);
	static_assert(offsetof(PluginVersionData, padding) == 0x308);
	static_assert(offsetof(PluginVersionData, compatibleVersions) == 0x30C);
	static_assert(offsetof(PluginVersionData, xseMinimum) == 0x34C);
	static_assert(sizeof(PluginVersionData) == 0x350);
```

File: include/REL/Relocation.h [hunk 2]
```diff
@@ -301,6 +301,15 @@ namespace REL
...
 		}
 
+		[[nodiscard]] constexpr std::uint32_t packed() const
+		{
+			auto major = std::get<0>(_impl);
+			auto minor = std::get<1>(_impl);
+			auto build = std::get<2>(_impl);
+			auto sub = std::get<3>(_impl);
+			return ((((major)&0xFF) << 24) | (((minor)&0xFF) << 16) | (((build)&0xFFF) << 4) | ((sub)&0xF));
+		}
```
Comment (Ryan-rsm-McKenzie) [inline include/REL/Relocation.h:311]:
```suggestion
		[[nodiscard]] constexpr std::uint32_t pack() const noexcept
		{
			return static_cast<std::uint32_t>(
				(_impl[0] & 0x0FF) << 24u |
				(_impl[1] & 0x0FF) << 16u |
				(_impl[2] & 0xFFF) << 4u |
				(_impl[3] & 0x00F) << 0u);
		}
```

File: src/SKSE/API.cpp [hunk 3]
```diff
@@ -75,8 +75,10 @@ namespace SKSE
 			stl::report_and_fail("interface is null"sv);
 		}
 
+#if 0
 		(void)REL::Module::get();
 		(void)REL::IDDatabase::get();
+#endif
```
Comment (Ryan-rsm-McKenzie) [inline src/SKSE/API.cpp:81]:
```suggestion
		(void)REL::Module::get();
		//(void)REL::IDDatabase::get();
```

**Review by Ryan-rsm-McKenzie — CHANGES_REQUESTED** [review]

**Comment by Ryan-rsm-McKenzie** [comment]
I would comment out the associated function for any offsets that weren't located

**Comment by Ryan-rsm-McKenzie** [comment]
I would use `#if 0` for commenting out multiple lines

**Review by Ryan-rsm-McKenzie — APPROVED** [review]

**Pull request merged**